### ISSUE TITLE
Collect: help

### ISSUE THREAD

**Issue by shadowspawn** [tj/commander.js#1225]
Collecting help issues that have come up multiple times.

Override or disable help option:

- <strike>#242 Make auto help optional</strike>
- <strike>#542 Hiding auto generated help options?</strike>
- #689 Don't Show options when there are no options
- <strike>#893 Override --help?</strike>
- <strike>#1215 Prevent/override automatic help generation</strike>
- #1370 Add method to get help option description

Struck out items included in Command v6.1.0

**Comment by shadowspawn** [comment]
Customisation, especially prefix/banner:

- #46 Add Support For Title?
- #578 Option to prefix help with text
- #1023 Customize help message for a command

**Comment by shadowspawn** [comment]
For customisation (e.g. prefix/banner), an implementation idea is:

Currently there is a single event which is emitted after the help is displayed, `--help`.

A possible approach for allowing more control might be to emit three events:

- `help:header`
- `help:body`
- `help:footer`

If there is a listener for `help:body` then it is called _instead_ of the normal help.

**Comment by DevNico** [comment]
Would be great to have the ability to hide options from the help command. eg. I have three commands:
- server:setup
- server:start
- server:stop

Now those commands all have the option -v to enable verbose output. (I know that I could just make this a global option but in some cases I don't want to do that) In this case in the general help command the -v option is shown three times. It would be great if I could just add the option in a way so it only shows up in the help for that specific command. Is this already somehow possible?  If not I would really like this as a feature

**Comment by shadowspawn** [comment]
Expanding the events to include the "group" of a command and its subcommands, which is effectively global if listened to on the root command:

1. `preGroupHelp`
2. `preHelp`
3. `help` | `groupHelp`
4. `postHelp`
5. `postGroupHelp`

**Comment by shadowspawn** [comment]
A possible syntax for disabling the built-in help option could be:

```js
program.helpOption(false);
```

with some symmetry with:

```js
program.addHelpCommand(false);
```

**Comment by shadowspawn** [comment]
I didn't like how the event listeners worked out as felt too complicated, and currently trying more direct api:

```js
program.addHelpText('after', 'See web site for further help');
```
```

**Comment by shadowspawn** [comment]
Trying `.addHelpText()` which takes a position as first parameter, and a string to add or a function returning a string. Position is 'before' or 'after' to affect just this command, and 'beforeAll' or 'afterAll' to affect this command and all its subcommands. 

`.help()` and `.outputHelp()` can be passed `{ error: true }` to write to stderr.

For example:

```js
program.addHelpText('afterAll', '/nSee the web site for more information');
```

See #1296 for more info

**Comment by shadowspawn** [comment]
Major improvements to custom help included in v7.0.0 pre-release #1386

**Comment by shadowspawn** [comment]
Commander 7.0.0 has shipped. Most of the collected issues have now been addressed.
